src/llmcompressor/pipelines/layer_sequential/helpers.py

```diff
@@ -0,0 +1,100 @@
+import contextlib
+import inspect
+from dataclasses import dataclass
+from typing import Any, Dict, List, Tuple
+
+import torch
+import tqdm
+from compressed_tensors.quantization import find_name_or_class_matches
+from compressed_tensors.utils import get_execution_device
+from torch.nn import Module
+from torch.utils.data.dataloader import DataLoader
+
+from llmcompressor.modifiers.utils.pytorch_helpers import apply_pad_mask_to_batch
+from llmcompressor.pipelines.cache import IntermediatesCache
+from llmcompressor.pytorch.utils.helpers import tensors_to_device
+from llmcompressor.utils.helpers import calibration_forward_context
+
+__all__ = ["match_modules", "capture_first_layer_intermediates", "to_next_layer_kwargs"]
+
+
+def match_modules(model: Module, target_names: List[str]) -> List[Module]:
+    names_layers = [
+        (name, module)
+        for name, module in model.named_modules()
+        if find_name_or_class_matches(name, module, target_names)
+    ]
+
+    names_layers = sorted(names_layers, key=lambda name_layer: name_layer[0])
+    return [layer for _name, layer in names_layers]
+
+
+def capture_first_layer_intermediates(
+    model: Module,
+    layers: List[Module],
+    dataloader: DataLoader,
+    mask_padding: bool = True,
+) -> IntermediatesCache:
+    model_device = get_execution_device(model)
+    intermediates = IntermediatesCache.empty(len(dataloader), torch.device("cpu"))
+    first_layer = layers[0]
+    signature = inspect.signature(first_layer.forward)
+
+    with calibration_forward_context(model), early_stop_hook(first_layer):
+        desc = "Preparing intermediates cache"
+        for batch_index, batch in enumerate(tqdm.tqdm(dataloader, desc=desc)):
+            batch = apply_pad_mask_to_batch(batch) if mask_padding else batch
+            batch = tensors_to_device(batch, model_device)
+
+            try:
+                model(**batch)
```

Comment:
If we're using multiple batches, i.e batch_size < dataset_size, the early stop would execute and return an intermediate output after the first batch from a layer which has not yet been quantized, correct? As _update_size by default is the entire dataset?

Reply:
If I'm reading your explanation correctly, then in most cases no.

If a modifier targets modules which execute before the first "layer" (for example, torch.nn.Embedding), then yes, that module would pass on outputs which do not propagate that module's compression error.

However, in most cases, we make the assumption that this is not the case.

We could fix this by running two forward passes (one calibration, one propagation) similar to the other layers, but this comes at considerable performance cost and is likely overkill for a non-default pipeline anyways.

Reply:
Thanks for the catch